src/github.com/matrix-org/dendrite/clientapi/readers/profile.go

```diff
@@ -175,13 +238,96 @@ func SetDisplayName(
 	}
 }
 
-func getLocalPart(userID string) string {
-	if !strings.HasPrefix(userID, "@") {
-		panic(fmt.Errorf("Invalid user ID"))
+func buildMembershipEvents(
+	memberships []authtypes.Membership, db *accounts.Database,
+	oldProfile *authtypes.Profile, changedKey string, newValue string,
+	userID string, cfg *config.Dendrite, queryAPI api.RoomserverQueryAPI,
+) ([]gomatrixserverlib.Event, error) {
+	ev := []gomatrixserverlib.Event{}
+
+	for _, membership := range memberships {
+		prevContent := events.MemberContent{
+			Membership:  "join",
+			DisplayName: oldProfile.DisplayName,
+			AvatarURL:   oldProfile.AvatarURL,
+		}
+
+		prev := prevMembership{
+			UserID:      userID,
+			PrevID:      membership.EventID,
+			PrevContent: prevContent,
+		}
+
+		builder := gomatrixserverlib.EventBuilder{
+			Sender:   userID,
+			RoomID:   membership.RoomID,
+			Type:     "m.room.member",
+			StateKey: &userID,
+		}
+
+		if err := builder.SetUnsigned(prev); err != nil {
+			return nil, err
+		}
+
+		content := events.MemberContent{
+			Membership: "join",
+		}
+
+		if changedKey == "displayname" {
+			content.DisplayName = newValue
+			content.AvatarURL = oldProfile.AvatarURL
+		} else if changedKey == "avatar_url" {
+			content.DisplayName = oldProfile.DisplayName
+			content.AvatarURL = newValue
+		}
+
+		if err := builder.SetContent(content); err != nil {
+			return nil, err
+		}
+
+		eventsNeeded, err := gomatrixserverlib.StateNeededForEventBuilder(&builder)
+		if err != nil {
+			return nil, err
+		}
+
+		// Ask the roomserver for information about this room
+		queryReq := api.QueryLatestEventsAndStateRequest{
+			RoomID:       membership.RoomID,
+			StateToFetch: eventsNeeded.Tuples(),
+		}
+		var queryRes api.QueryLatestEventsAndStateResponse
+		if queryErr := queryAPI.QueryLatestEventsAndState(&queryReq, &queryRes); queryErr != nil {
+			return nil, err
+		}
+
+		authEvents := gomatrixserverlib.NewAuthEvents(nil)
+
+		// Iterating the old way because range seems to mess things up. Might be
+		// worth investigating.
+		for i := 0; i < len(queryRes.StateEvents); i++ {
+			authEvents.AddEvent(&queryRes.StateEvents[i])
+		}
```

Comment:
Ah yes this can be a bit subtle.

This should work:

```go
for i := range(queryRes.StateEvents) {
    authEvents.AddEvent(&queryRes.StateEvents[i])
}
```

However this will have problems:

```go
for _, event := range(queryRes.StateEvents) {
    authEvents.AddEvent(&event)
}
````

The reason is that authEvents will be passed an address to the iteration variable "events" that is clobbered on each iteration of the loop.